src/mscorlib/shared/System/Double.cs

```diff
@@ -221,16 +223,18 @@ public bool Equals(Double obj)
         //The hashcode for a double is the absolute value of the integer representation
         //of that double.
         //
+        [MethodImpl(MethodImplOptions.AggressiveInlining)] // 64-bit constants make the IL unusually large that makes the inliner to reject the method
         public unsafe override int GetHashCode()
         {
-            double d = m_value;
-            if (d == 0)
+            var bits = Unsafe.As<double, long>(ref m_value);
+
+            if (((bits - 1) & 0x7FFFFFFFFFFFFFFF) >= 0x7FF0000000000000)
```

Comment:
Why `(bits - 1)`?

Reply:
This definitely needs a comment.

At first glance, it looks like you are grabbing all values matching (IsInfinity || IsNaN), but this is actually doing (IsNaN || IsZero).

Reply:
Comment added